src/asyncresults.jl

```diff
@@ -0,0 +1,242 @@
+"An asynchronous PostgreSQL query"
+mutable struct AsyncResult
+    "The LibPQ.jl Connection used for the query"
+    jl_conn::Connection
+
+    "Whether this AsyncResult has locked the Connection"
+    conn_locked::Atomic{Bool}
+
+    "Keyword arguments to pass to Result on creation"
+    result_kwargs::Ref
+
+    "Task which errors or returns a LibPQ.jl Result which is created once available"
+    result_task::Task
+
+    function AsyncResult(jl_conn::Connection, conn_locked::Atomic{Bool}, result_kwargs::Ref)
+        async_result = new(jl_conn, conn_locked, result_kwargs)
+        jl_conn.async_result = async_result
+        return async_result
+    end
+end
+
+function AsyncResult(jl_conn::Connection, conn_locked::Atomic{Bool}; kwargs...)
+    AsyncResult(jl_conn, conn_locked, Ref(kwargs))
+end
+
+function AsyncResult(jl_conn::Connection; kwargs...)
+    AsyncResult(jl_conn, Atomic{Bool}(true); kwargs...)
+end
+
+function Base.show(io::IO, async_result::AsyncResult)
+    status = if isready(async_result)
+        if iserror(async_result)
+            "errored"
+        else
+            "finished"
+        end
+    else
+        "in progress"
+    end
+    print(io, typeof(async_result), " (", status, ")")
+end
+
+function _ensure_unlocked(async_result::AsyncResult)
+    was_locked = atomic_xchg!(async_result.conn_locked, false)
+    if was_locked
+        async_result.jl_conn.async_result = nothing
+        unlock(async_result.jl_conn)
+    end
+    return was_locked
+end
+
+function handle_result(async_result::AsyncResult, success::Bool; throw_error=true)
+    log_function = throw_error ? error : warn
+
+    if success
+        async_result.result_task = consume(async_result; throw_error=throw_error)
+    else
+        msg = error_message(async_result.jl_conn)
+        _ensure_unlocked(async_result)
+        log_function(LOGGER, msg)
+    end
+
+    return async_result
+end
+
+"""
+    consume(async_result::AsyncResult; throw_error=true) -> Task
+
+Run a task which executes the query in `async_result` and waits for results.
+
+This implements the loop described in the PostgreSQL documentation for
+[Asynchronous Command Processing](https://www.postgresql.org/docs/10/libpq-async.html).
+
+The `throw_error` option only determines whether to throw errors when handling the new
+[`Result`](@ref)s; the `Task` may error for other reasons related to processing the
+asynchronous loop.
+
+The result returned from the `Task` will be the [`Result`](@ref) of the last query run (the
+only query if using parameters).
+Any errors produced by the queries will be thrown together in a `CompositeException` by
+`@sync`.
+"""
+function consume(async_result::AsyncResult; throw_error=true)
+    @async begin
```

Comment:
Should we be checking (as an assertion) we own the lock on the connection at the start of this function?
The one we free at the end
I assume we aquired it somewhere earlier before calling this?

Reply:
Yeah that's a good idea.